src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -97,8 +100,20 @@ class Executor {
                 : Status(StatusCode(reply.status().code()), reply.status().message()); \
         callback(status, reply);                                                       \
         delete executor;                                                               \
+      } else if (gcs_is_down_) {                                                       \
+        pending_requests_.emplace_back(executor);                                      \
+        if (pending_requests_.size() >                                                 \
+            ::RayConfig::instance().gcs_grpc_max_request_queued()) {                   \
+          RAY_LOG(WARNING) << "Pending queue for failed GCS request has reached the "  \
+                           << "limit. Blocking the current thread until GCS is back";  \
+          while (gcs_is_down_ && !shutdown_) {                                         \
+            CheckChannelStatus();                                                      \
+            std::this_thread::sleep_for(std::chrono::milliseconds(                     \
```

Comment:
Is it acceptable to block until GCS is up? How about using a timer or something else and run `CheckChannelStatus()` asynchronously `gcs_client_check_connection_status_interval_milliseconds` later?

Reply:
I think it's not acceptable, but it's the best we can do for now.

The concern here is that there are too many requests queued when GCS is down. And when this happens we either:
- fail some request 
- blocking until GCS is up so it can clear the queue.

The former one is better, but given that our codebase is not handling this very well and a lot of places assume GCS request is always successful, it'll need more effort. I guess the long-term plan is to make sure gRPC error is handled correctly. Some of this will be handled in GCS HA project, others might need some cleanup or redesign of the protocol.


Actually, we do the check async if the queue is not full.